gnuplot/src/color.rs

```diff
@@ -0,0 +1,413 @@
+	///# let x = [1,2,3];
+	///# let y = [4,5,6];
+	/// assert!(frac >= 0.0);
+	/// assert!(frac <= 1.0);
+	///
+	/// let mut fg = Figure::new();
+	/// let ax = fg.axes2d();
+	/// ax.set_cb_range(Fix(min), Fix(max));
+	/// let col1 = Color(PaletteFracColor(frac));
+	/// let cb_range = max - min;
+	/// let col2 = Color(PaletteCBColor(min + (frac * cb_range)));
+	/// ax.lines(x, y, &[col1]);
+	/// ax.lines(x, y, &[col2]);
+	/// ```
+	/// the two lines should give the same color for any values of `max` and `min`, and `0 <= frac <= 1`.
+	PaletteCBColor(f64),
+	/// Vector of `f64` values which act as indexes into the current palette to set the color of
+	/// each data point. These variable values work in the same was as the single fixed value supplied
+	/// to a [PaletteCBColor]
+	VariablePaletteColor(Vec<f64>),
+	/// Similar to `VariablePaletteColor` in that it takes a `Vec<f64>` to set the indexes into the
+	/// color map for each data point, but in addition to the color data it takes a string hold the name
+	/// of the color map to use. This should have been previously created in the workspace using the
+	/// [create_colormap()](crate::AxesCommon::create_colormap) function.
+	SavedColorMap(T, Vec<f64>),
+	/// Set the color of all elements of the plot to the `n`th color in the current gnuplot color cycle.
+	Index(ColorIndex),
+	/// A color type that sets the color per element using a index `n` which represents the `n`th
+	/// color in the current gnuplot color scheme. In gnuplot this is the last element in the plot command,
+	/// in Rust gnuplot, the color type takes a vector of u8, where each index is treated the same as the
+	/// fixed `IndexColor`.
+	/// This is useful for setting bars/boxes etc to be
+	/// the same color from multiple plot commands. The `variable_color` example has examples of this usage.
+	VariableIndex(Vec<ColorIndex>),
+	/// Set the color of the plot to the current background color.
+	Background,
+	/// Fixed black color
+	Black,
+}
+
+impl<T: Display> ColorType<T>
+{
+	/// Returns the gnuplot string that will produce the requested color
+	pub fn command(&self) -> String
+	{
+		let str = match self
+		{
+			RGBString(s) => &format!(r#"rgb "{}""#, s),
+			RGBInteger(r, g, b) => &format!(r#"rgb {}"#, from_argb(0, *r, *g, *b)),
+			ARGBInteger(a, r, g, b) => &format!(r#"rgb {}"#, from_argb(*a, *r, *g, *b)),
+			VariableRGBInteger(_) => "rgb variable",
+			VariableARGBInteger(_) => "rgb variable",
+			PaletteFracColor(v) => &format!("palette frac {v}"),
+			PaletteCBColor(v) => &format!("palette cb {v}"),
+			VariablePaletteColor(_) => "palette z",
+			SavedColorMap(s, _) => &format!("palette {s}"),
+			VariableIndex(_) => "variable",
+			Background => "background",
+			Index(n) => &format!("{}", n),
+			Black => "black",
+		};
+		String::from(str)
+	}
+
+	pub fn data(&self) -> Vec<f64>
+	{
+		match self
+		{
+			RGBString(_) => panic!("data() called on non-variable color type."),
```

Comment:
Perhaps handle the non-data containing variants with a wildcard pattern? If you are concerned about making sure there's errors if new variants are added, I'd still prefer using a OR pattern, e.g.

```rust
RGBString(_) | PaletteFracColor(_)
```

and so on. Also, the panic should mention what variant was erroneously used.

Reply:
Agree a wildcard pattern is better. As this is an crucial implementation step, I would expect anyone adding more variable colors to need to update this without a compilation error. 

The current approach is the form I used when implementing (to make sure all branches covered). Better a wildcard now.